### ISSUE TITLE
Deteriorated IDE experience for python (autocomplete)

### ISSUE THREAD

**Issue by komalali** [pulumi/pulumi#7367]
We recently made a change, specifically https://github.com/pulumi/pulumi/pull/6827 that has led to a bit of a deteriorated IDE experience. We also don't have `class` docstrings, only `__init__` docstrings, which causes a bit of a weird experience in VSCode. The issues are detailed below.

#### Module Autocomplete

If we `import pulumi_aws as aws`, autocomplete appears to work fine.

<img width="600" alt="Screen Shot 2021-06-25 at 11 21 18 AM" src="https://user-images.githubusercontent.com/24326455/123455955-e9945800-d5a7-11eb-97bf-73b741e9c00a.png">
<img width="652" alt="Screen Shot 2021-06-25 at 11 21 27 AM" src="https://user-images.githubusercontent.com/24326455/123455961-eac58500-d5a7-11eb-9dfb-142f96298cb3.png">
<img width="660" alt="Screen Shot 2021-06-25 at 11 21 50 AM" src="https://user-images.githubusercontent.com/24326455/123455969-ec8f4880-d5a7-11eb-8c66-9028bbc21ab4.png">

On the other hand, if we do `from pulumi_aws import iam`, we get the initial autocomplete for `iam`, but no further autocomplete for the module itself.

<img width="725" alt="Screen Shot 2021-06-25 at 11 20 42 AM" src="https://user-images.githubusercontent.com/24326455/123456038-003aaf00-d5a8-11eb-8f58-b7edf421b2f9.png">
<img width="594" alt="Screen Shot 2021-06-25 at 11 20 59 AM" src="https://user-images.githubusercontent.com/24326455/123456055-06309000-d5a8-11eb-9ae2-1525cb3b59a4.png">

This behavior seems related to the changes in https://github.com/pulumi/pulumi/pull/6827

##### Module Autocomplete

If we `import pulumi_aws as aws`, we don't get any autocomplete for the modules, only the direct exports from the parent module. Again, I would attribute this to the changes in https://github.com/pulumi/pulumi/pull/6827

<img width="712" alt="Screen Shot 2021-06-25 at 11 28 25 AM" src="https://user-images.githubusercontent.com/24326455/123464050-eef6a000-d5b1-11eb-9fb6-6f9f66fc7c3d.png">

If we continue on, we do get autocomplete for the exports in `iam` itself.
<img width="732" alt="Screen Shot 2021-06-25 at 12 37 17 PM" src="https://user-images.githubusercontent.com/24326455/123464260-3715c280-d5b2-11eb-9194-d4b2d9dc5fa2.png">

This also works if we import `from pulumi_aws import iam`
<img width="702" alt="Screen Shot 2021-06-25 at 12 37 35 PM" src="https://user-images.githubusercontent.com/24326455/123464305-4432b180-d5b2-11eb-8b6d-03ecf6f8b5ba.png">

#### Class docstrings

Since we only emit docstrings for the `__init__` functions, we don't see the documentation when hovering over the class:

<img width="545" alt="Screen Shot 2021-06-25 at 12 55 14 PM" src="https://user-images.githubusercontent.com/24326455/123466231-bad0ae80-d5b4-11eb-94a9-96da4c06986e.png">
<img width="516" alt="Screen Shot 2021-06-25 at 12 51 01 PM" src="https://user-images.githubusercontent.com/24326455/123465695-12224f00-d5b4-11eb-8836-12ac7f8f1cbc.png">

But we do see it when writing the constructor:

<img width="821" alt="Screen Shot 2021-06-25 at 12 56 44 PM" src="https://user-images.githubusercontent.com/24326455/123466317-d1770580-d5b4-11eb-8bd3-8993bf6cdf22.png">
<img width="748" alt="Screen Shot 2021-06-25 at 12 51 18 PM" src="https://user-images.githubusercontent.com/24326455/123465739-1ea6a780-d5b4-11eb-906b-cbffa354c907.png">

For the above issue, we should consider emitting a `class` docstring as well as the `__init__` docstring so the information is available in either place.

**Comment by t0yv0** [comment]
Thanks for the excellent bug report, it is interesting that the IDEs diverge on this.

A breaking solution would be the cleanest option here; we would remove sub-module variables completely. That is, this would break:

```
import pulumi_aws as aws
x = aws.ec2.X()
```

And would require rewriting to:

```
import pulumi_aws.ec2 as ec2
```

With this solution, lazy imports are not needed, and the codebase becomes a "normal" codebase, without perf degradation.

Other options I can think of:

- Finding a type-preserving solution to lazy_import (working both in mypy and pylance/whatever engines the IDEs use) - may be very tricky

- Importing frequently used submodules normally; that is, balancing perf/IDE correctness on basis of usage; only using lazy loading for infrequently used modules

- Detecting environment difference between IDE usage and Pulumi usage; I think this probably is best done by Pulumi providing an environment variable when running Python commands. Currently we discriminate by `if typing.TYPE_CHECKING` to decide whether to load modules normally or lazily. This makes mypy work but unfortunately is not respected by IDEs. If we could detect the "we're in the IDE situation", we could extend this condition, so that IDE would get normal imports, but pulumi programs would get lazy imports and speedup.

**Comment by t0yv0** [comment]
I think I like option (3). I would like to maybe do this item as part of quality sprint.

**Comment by komalali** [comment]
Hmm, interesting, I kinda like that idea, but I'm not so sure I feel great about pushing the onus onto the user to have to set an env var to get autocomplete to work. Is that what you're suggesting here?

As a general rule, I think code discoverability is a minimum bar that we should try to always prioritize as the providers of these libraries. 

I was talking to @lukehoban about this during our 1:1, and we were wondering whether the lazy loading is:
1) actually just fixing a symptom of a deeper issue (what about loading modules takes so long?)
2) worth the gains in non-azure-native situations since azure-native is the outlier by far.

**Comment by komalali** [comment]
I was thinking of an alternative, i.e. check if "we're in a pulumi situation" and only lazy-load in that situation. We could set the env var to detect it in https://github.com/pulumi/pulumi/blob/480173a57f7360c4f261c1b28efbb02cd4381bb5/sdk/python/cmd/pulumi-language-python-exec 
However, this would not work for automation-api/multi-lang scenarios where there are alternative entry-points.

**Comment by t0yv0** [comment]
Not necessarily - it would be transparent to the user. Pulumi CLI would set this env var when calling Python programs such as main.py as an optimization.

**Comment by komalali** [comment]
>Pulumi CLI would set this env var when calling Python programs such as main.py as an optimization.

Ah I see, so it would only work in Pulumi CLI-driven programs (i.e. excluding automation API and maybe multilang scenarios)

**Comment by t0yv0** [comment]
Right, sounds like we're thinking along the similar path. So what would be the alternative entry-points, do we automation-api/multi-lang invoke python not from a subprocess of pulumi CLI? Can do some research when doing this ticket.

**Comment by t0yv0** [comment]
Hm, I guess it would be possible. If any user code would import pulumi_aws to run outside of pulumi CLI, it might hit the slow import path.

**Comment by komalali** [comment]
Yeah, automation api starts with the user running `python main.py`. For local programs, it still uses the CLI to actually invoke the program, but inline programs are defined within the script that is running with `python main.py` so the script itself is the entrypoint and therefore the modules are loaded when the script runs.

**Comment by t0yv0** [comment]
Regarding this:

```
I was talking to @lukehoban about this during our 1:1, and we were wondering whether the lazy loading is:

actually just fixing a symptom of a deeper issue (what about loading modules takes so long?)
worth the gains in non-azure-native situations since azure-native is the outlier by far.
```

I think that the underlying issue is that force-loading sub-modules from parent module __init__.py is not a great idea, this defeats how Python module loading works, and forces it to load everything even if your code does not use it. Removing that would get us to the best of all worlds here. However it is a breaking change.

**Comment by komalali** [comment]
>I think that the underlying issue is that force-loading sub-modules from parent module init.py is not a great idea, this defeats how Python module loading works, and forces it to load everything even if your code does not use it.

Yeah, this is indeed an unfortunate situation we have found ourselves in.

**Comment by t0yv0** [comment]
I'll play some more with this. Maybe there's a trivial way to detect when we're in the IDE typechecking context.

**Comment by 1oglop1** [comment]
Adding a reference to the PyCharm issue not being able to read from `lazy_import`s https://youtrack.jetbrains.com/issue/PY-49357

**Comment by 1oglop1** [comment]
Adding few more examples

```python
from pulumi_aws import s3
a_1 = s3.BucketObject

a_2 = s3.BucketObject()

import pulumi_aws as aws
b_2 = aws.s3.BucketObject
b_1 = aws.s3.BucketObject()

import pulumi_aws.s3 as s3_again
d = s3_again.BucketObject
c = s3_again.BucketObject()
```
PyCharm `a_1`
![image](https://user-images.githubusercontent.com/7480694/123584114-12ebf880-d7e1-11eb-8955-9eef11e7909d.png)

PyCharm `a_2`
![image](https://user-images.githubusercontent.com/7480694/123584321-7413cc00-d7e1-11eb-810d-44016e34fdb9.png)

PyCharm `b_2`
![image](https://user-images.githubusercontent.com/7480694/123584352-83931500-d7e1-11eb-9fee-350e559097e7.png)

PyCharm `b_1`
![image](https://user-images.githubusercontent.com/7480694/123584388-9279c780-d7e1-11eb-9b02-9b6ed20df671.png)

PyCharm `d`
![image](https://user-images.githubusercontent.com/7480694/123584423-a1607a00-d7e1-11eb-9656-7e943db8c6fd.png)

PyCharm `c`
![image](https://user-images.githubusercontent.com/7480694/123584472-b50be080-d7e1-11eb-979e-253533a50474.png)

**Comment by komalali** [comment]
@t0yv0 typing stub files might be a viable option to solve our autocomplete woes as @justinvp mentioned https://github.com/pulumi/pulumi/issues/5582#issuecomment-868546209

I'm using a similar strategy for the config getters... https://github.com/pulumi/pulumi/pull/7447/files#diff-95e293463996b4b9eb1693141340fb0b6ffb4a518ea17242239aec3532c189f5

(It's a work in progress)

**Comment by t0yv0** [comment]
That sounds a bit painful no? Maybe I don't fully understand. But here is how it appears to me:

- we already emit type information the .py files
- if we modified codegen to not emit the info in .py files and instead emit it in .pyi files it might work
- however I looked and did not find a way to "symlink" a variable type information, that is, declare that variable `config` has the same type structure as module "config"
- without symlinking we'd have to duplicate the type information emission which bloats the package

I still want to play with lighter weight fixes here a bit before going in that direciton.

**Comment by komalali** [comment]
>however I looked and did not find a way to "symlink" a variable type information, that is, declare that variable config has the same type structure as module "config"

I'm not really following here, but yeah, totally do your thing. As long as the end goal of having autocomplete and type hints function correctly I'm not really married to any specific fix.

**Comment by justinvp** [comment]
The `config` module in our provider SDKs aren't typically very large, so I'm personally not too concerned about adding a single stub file for it, if it provides the IDE experience we want along with the runtime semantics we want in the implementation.

BTW, I'm not suggesting we create stub files for _everything_. Aside from `config`, I was just wondering if we could use a single stub file (just `__init__.pyi`) for each top-level module that has submodules, that makes it look like the submodules are imported greedily, rather than lazily, so that IDEs can provide code completion for the submodules, since the IDEs don't appear to respect the [`if typing.TYPE_CHECKING:` we added for mypy](https://github.com/pulumi/pulumi-aws/blob/3e04177ade1de9b326346c63c5553eb7bac11f6c/sdk/python/pulumi_aws/__init__.py#L30).

Granted, I have not tried this myself, so I don't know if we can get away with just `__init__.pyi` for each module that has lazy submodules. If this solution requires having stub files for everything (e.g. `s3/bucket.pyi`) I would also be concerned about bloat.

But if there are other ways we can address the problem, I'm all for it.

**Comment by t0yv0** [comment]
That's a good idea, I'm going to try that again with `__init__.pyi`, I have a vague memory it didn't work but need a refresher.

**Comment by wannessels** [comment]
Any idea why Pycharm infers type `Optional[Any]` for all parameters (except resource_name and opts)?

**Comment by justinvp** [comment]
> Any idea why Pycharm infers type `Optional[Any]` for all parameters (except resource_name and opts)?

@wannessels, unfortunately, this is a bug in PyCharm: https://youtrack.jetbrains.com/issue/PY-29257

**Comment by wannessels** [comment]
> @wannessels, unfortunately, this is a bug in PyCharm: https://youtrack.jetbrains.com/issue/PY-29257

This annoyed me _a lot_, so I did some yak shaving over the weekend and [fixed it](https://github.com/JetBrains/intellij-community/pull/1655) :-)

I now get proper type hints for all parameters:
![image](https://user-images.githubusercontent.com/1406987/128648258-d2bcabc8-a743-4bfb-a2e6-9d95a8a5b089.png)
The `InputType`s are incredibly long, but it's usable at least.
